include/ginkgo/core/preconditioner/schwarz.hpp

```diff
@@ -0,0 +1,268 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_PRECONDITIONER_SCHWARZ_HPP_
+#define GKO_PUBLIC_CORE_PRECONDITIONER_SCHWARZ_HPP_
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+#include <ginkgo/core/multigrid/amgx_pgm.hpp>
+#include <ginkgo/core/multigrid/multigrid_level.hpp>
+
+
+namespace gko {
+/**
+ * @brief The Preconditioner namespace.
+ *
+ * @ingroup precond
+ */
+namespace preconditioner {
+
+
+/**
+ * A Schwarz preconditioner is a simple domain decomposition preconditioner that
+ * generalizes the Block Jacobi preconditioner, incorporating options for
+ * different local subdomain solvers and overlaps between the subdomains.
+ *
+ * See Iterative Methods for Sparse Linear Systems (Y. Saad) for a general
+ * treatment and variations of the method.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  integral type used to store pointers to the start of each
+ *                    block
+ *
+ * @ingroup schwarz
+ * @ingroup precond
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Schwarz : public EnableLinOp<Schwarz<ValueType, IndexType>>,
+                public WritableToMatrixData<ValueType, IndexType>,
+                public Transposable {
+    friend class EnableLinOp<Schwarz>;
+    friend class EnablePolymorphicObject<Schwarz, LinOp>;
+
+public:
+    using EnableLinOp<Schwarz>::convert_to;
+    using EnableLinOp<Schwarz>::move_to;
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using mat_data = matrix_data<ValueType, IndexType>;
+    using transposed_type = Schwarz<ValueType, IndexType>;
+
+    /**
+     * Returns the number of blocks of the operator.
+     *
+     * @return the number of blocks of the operator
+     */
+    size_type get_num_subdomains() const noexcept { return num_subdomains_; }
+
+    void write(mat_data& data) const override;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    /**
+     * Returns the subdomain matrices.
+     *
+     * @return the subdomain matrices
+     */
+    std::vector<std::shared_ptr<LinOp>> get_subdomain_matrices() const
+    {
+        return subdomain_matrices_;
+    }
+
+    /**
+     * A struct that stores the coarse operator and solver and generates the
+     * coarse solver with the given operator
+     */
+    struct coarse_solver_type {
```

Comment:
Why is that necessary? Can't you use a `Multigrid` solver with 2 levels?

Reply:
No, a multigrid is different because it restricts the error and only does a solve on the coarse grid. What I want to possibly do is solve on each level. So, what I need is the functionality of MultigridLevel (which stores the R, A and P) and a LinOpFactory for each level, so I think this struct combines both and generates the coarse solvers from the `MultigridLevel`s.

Reply:
What do you mean by solve on each level? Why is that different from setting the coarsest level to the one you are doing the solve on, and just leave out the rest?